snuba/clickhouse/optimize/optimize.py

```diff
@@ -330,6 +385,80 @@ def optimize_partition_runner(
             return
 
 
+def _add_optimize_jobs_to_futures(
+    executor: ThreadPoolExecutor,
+    futures: List[Future[None]],
+    active_futures: set[Future[Any]],
+    partitions_to_optimize: deque[str],
+    configured_num_threads: int,
+    schedule: OptimizationSchedule,
+    clickhouse: ClickhousePool,
+    database: str,
+    table: str,
+    tracker: OptimizedPartitionTracker,
+    clickhouse_host: str,
+) -> None:
+    for _ in range(configured_num_threads - len(active_futures)):
+        if not partitions_to_optimize:
+            break
+
+        futures.append(
+            executor.submit(
+                optimize_partitions,
+                clickhouse,
+                database,
+                table,
+                [partitions_to_optimize.popleft()],
+                schedule.cutoff_time,
+                tracker,
+                clickhouse_host,
+            )
+        )
+
+
+def _jitter_start_time(start_jitter: Optional[int] = None) -> None:
+    if start_jitter is not None:
+        logger.info(
+            f"{threading.current_thread().name}: Jittering start time by"
+            f" {start_jitter} minutes"
+        )
+        time.sleep(start_jitter * 60)
+
+
+def _set_up_futures_if_initial_run(
+    executor: ThreadPoolExecutor,
+    futures: List[Future[None]],
+    active_futures: set[Future[Any]],
+    partitions_to_optimize: deque[str],
+    configured_num_threads: int,
+    schedule: OptimizationSchedule,
+    clickhouse: ClickhousePool,
+    database: str,
+    table: str,
+    tracker: OptimizedPartitionTracker,
+    clickhouse_host: str,
+    start_time_jitter_for_each_partition: Sequence[int],
+    is_initial_run: bool,
+) -> None:
```

Comment:
I'm not sure if splitting this code out into a separate function makes things more readable, because there's an awful lot of parameters that need to be passed. Also, this function is a no-op when `is_initial_run` is `False`, so even if we want to keep this function separate we should get rid of this parameter and just not call this method outside of the initial run. I think we can easily do this by moving the call outside of the body of the `while remaining_partitions:` loop.

This is what I'm thinking, in psuedo-code:
```
with ThreadPoolExecutor(max_workers=32) as executor:
    futures = []
    # jitter start time
    for jitter in start_time_jitter_for_each_partition:
        futures.append(executor.submit(_jitter_start_time, jitter))

    partitions_to_optimize = deque(partitions)
    while partitions_to_optimize:
        while len(futures) < configured_num_threads:
            futures.append(executor.submit(_new_optimize_job(partitions_to_optimize.popleft())))
        _, futures = concurrent.futures.wait(futures, return_when=concurrent.futures.FIRST_COMPLETED)
```
Thoughts?